// pg-boss/dist/boss.js
import EventEmitter from 'node:events';
import * as plans from "./plans.js";
import { delay, unwrapSQLResult } from "./tools.js";
import * as types from "./types.js";
import { emitAndPersistWarning } from "./warning.js";
const events = {
    error: 'error',
    warning: 'warning'
};
// Default thresholds and warning messages
const WARNINGS = {
    SLOW_QUERY: { seconds: 30, message: 'Warning: slow query. Your queues and/or database server should be reviewed' },
    LARGE_QUEUE: { size: 10_000, message: 'Warning: large queue backlog. Your queue should be reviewed' }
};
const WARNING_TYPES = {
    SLOW_QUERY: 'slow_query',
    QUEUE_BACKLOG: 'queue_backlog'
};
class Boss extends EventEmitter {
    #stopped;
    #stopping;
    #maintaining;
    #superviseInterval;
    #db;
    #config;
    #manager;
    #slowQuerySeconds;
    #largeQueueSize;
    events = events;
    constructor(db, manager, config) {
        super();
        this.#db = db;
        this.#config = config;
        this.#manager = manager;
        this.#stopped = true;
        this.#stopping = false;
        this.#slowQuerySeconds = config.warningSlowQuerySeconds || WARNINGS.SLOW_QUERY.seconds;
        this.#largeQueueSize = config.warningQueueSize || WARNINGS.LARGE_QUEUE.size;
    }
    get maintaining() {
        return !!this.#maintaining;
    }
    async start() {
        if (this.#stopped) {
            this.#stopping = false;
            this.#superviseInterval = setInterval(() => this.#onSupervise(), this.#config.superviseIntervalSeconds * 1000);
            this.#stopped = false;
        }
    }
    async stop() {
        if (!this.#stopped) {
            this.#stopping = true;
            if (this.#superviseInterval)
                clearInterval(this.#superviseInterval);
            this.#stopped = true;
            while (this.#maintaining) {
                await delay(10);
            }
        }
    }
    get #warningContext() {
        return {
            emitter: this,
            db: this.#db,
            schema: this.#config.schema,
            persistWarnings: this.#config.persistWarnings,
            warningEvent: events.warning,
            errorEvent: events.error
        };
    }
    async #executeQuery(query) {
        if (typeof (query) === 'string') {
            query = { text: query, values: [] };
        }
        const started = Date.now();
        const result = unwrapSQLResult(await this.#db.executeSql(query.text, query.values));
        const elapsed = (Date.now() - started) / 1000;
        if (elapsed > this.#slowQuerySeconds ||
            this.#config.__test__warn_slow_query) {
            await emitAndPersistWarning(this.#warningContext, WARNING_TYPES.SLOW_QUERY, WARNINGS.SLOW_QUERY.message, { elapsed, sql: query.text, values: query.values });
        }
        return result;
    }
    async #onSupervise() {
        try {
            if (this.#stopped)
                return;
            if (this.#maintaining)
                return;
            if (this.#config.__test__throw_maint) {
                throw new Error(this.#config.__test__throw_maint);
            }
            this.#maintaining = true;
            if (this.#config.__test__delay_maint_ms) {
                await delay(this.#config.__test__delay_maint_ms);
            }
            const queues = await this.#manager.getQueues();
            !this.#stopped && (await this.supervise(queues));
        }
        catch (err) {
            this.emit(events.error, err);
        }
        finally {
            this.#maintaining = false;
        }
    }
    async #maintainWarnings() {
        if (!this.#config.persistWarnings || !this.#config.warningRetentionDays) {
            return;
        }
        const sql = plans.deleteOldWarnings(this.#config.schema, this.#config.warningRetentionDays);
        await this.#executeQuery(sql);
    }
    async #ensureQueueStatsPartitions() {
        const sql = plans.ensureQueueStatsPartitions(this.#config.schema);
        await this.#executeQuery(sql);
    }
    async #maintainQueueStats() {
        if (!this.#config.persistQueueStats || !this.#config.queueStatRetentionDays) {
            return;
        }
        const sql = this.#config.noTablePartitioning
            ? plans.deleteOldQueueStats(this.#config.schema, this.#config.queueStatRetentionDays)
            : plans.dropOldQueueStatsPartitions(this.#config.schema, this.#config.queueStatRetentionDays);
        await this.#executeQuery(sql);
    }
    async supervise(value) {
        let queues;
        if (Array.isArray(value)) {
            queues = value;
        }
        else {
            queues = await this.#manager.getQueues(value);
        }
        // Ensure today's/tomorrow's partitions exist before any insertQueueStats below. Retention
        // (#maintainWarnings/#maintainQueueStats) runs at the tail. Both live here, in the public
        // supervise() path that also performs the writes, rather than in the timer-only #onSupervise
        // wrapper — so manual supervise() callers (instances run with the built-in supervisor disabled)
        // get partitions provisioned and old data pruned, not just job retention.
        if (this.#config.persistQueueStats && !this.#config.noTablePartitioning && !this.#stopping) {
            await this.#ensureQueueStatsPartitions();
        }
        const queueGroups = queues.reduce((acc, q) => {
            const { table } = q;
            acc[table] = acc[table] || { table, queues: [] };
            acc[table].queues.push(q);
            return acc;
        }, {});
        const heartbeatQueueNames = new Set(queues.filter(q => q.heartbeatSeconds != null).map(q => q.name));
        for (const queueGroup of Object.values(queueGroups)) {
            if (this.#stopping)
                return;
            const { table, queues } = queueGroup;
            const names = queues.map((i) => i.name);
            while (names.length) {
                if (this.#stopping)
                    return;
                const chunk = names.splice(0, 100);
                await this.#monitor(table, chunk, heartbeatQueueNames);
                await this.#maintain(table, chunk);
            }
        }
        if (this.#stopping)
            return;
        await this.#maintainWarnings();
        await this.#maintainQueueStats();
    }
    async #monitor(table, names, heartbeatQueueNames) {
        if (this.#stopping)
            return;
        const command = plans.trySetQueueMonitorTime(this.#config.schema, names, this.#config.monitorIntervalSeconds);
        const { rows } = await this.#executeQuery(command);
        if (this.#stopping)
            return;
        if (rows.length) {
            const queues = rows.map((q) => q.name);
            const cacheStatsSql = plans.cacheQueueStats(this.#config.schema, table, queues, this.#config.noAdvisoryLocks);
            const { rows: rowsCacheStats } = await this.#executeQuery(cacheStatsSql);
            if (this.#config.persistQueueStats) {
                const insertSql = plans.insertQueueStats(this.#config.schema, queues, this.#config.noAdvisoryLocks);
                await this.#executeQuery(insertSql);
            }
            if (this.#stopping)
                return;
            // Coerce with Number(): CockroachDB returns these integer columns as strings, so a bare `>`
            // would compare lexicographically ("100" > "9" === false) and silently miss the backlog. On
            // standard Postgres these are already numbers, so Number() is a no-op.
            const warnings = rowsCacheStats.filter(i => Number(i.queuedCount) > (Number(i.warningQueueSize) || this.#largeQueueSize));
            for (const warning of warnings) {
                await emitAndPersistWarning(this.#warningContext, WARNING_TYPES.QUEUE_BACKLOG, WARNINGS.LARGE_QUEUE.message, warning);
            }
            // CockroachDB rejects the multi-mutation failJobs() CTE these use, so under noMultiMutationCte
            // route expiry through the manager's split select/delete/re-insert variants instead.
            if (this.#config.noMultiMutationCte) {
                await this.#manager.failJobsByTimeoutDistributed(table, queues);
            }
            else {
                const sql = plans.failJobsByTimeout(this.#config.schema, table, queues, this.#config.noAdvisoryLocks);
                await this.#executeQuery(sql);
            }
            if (this.#stopping)
                return;
            const heartbeatQueues = queues.filter(q => heartbeatQueueNames.has(q));
            if (heartbeatQueues.length) {
                if (this.#config.noMultiMutationCte) {
                    await this.#manager.failJobsByHeartbeatDistributed(table, heartbeatQueues);
                }
                else {
                    const heartbeatSql = plans.failJobsByHeartbeat(this.#config.schema, table, heartbeatQueues, this.#config.noAdvisoryLocks);
                    await this.#executeQuery(heartbeatSql);
                }
            }
        }
    }
    async #maintain(table, names) {
        if (this.#stopping)
            return;
        const command = plans.trySetQueueDeletionTime(this.#config.schema, names, this.#config.maintenanceIntervalSeconds);
        const { rows } = await this.#executeQuery(command);
        if (this.#stopping)
            return;
        if (rows.length) {
            const queues = rows.map((q) => q.name);
            const sql = plans.deletion(this.#config.schema, table, queues, this.#config.noAdvisoryLocks);
            await this.#executeQuery(sql);
            const depSql = plans.cleanupDependencies(this.#config.schema, table, queues, this.#config.noAdvisoryLocks);
            await this.#executeQuery(depSql);
        }
    }
}
export default Boss;
