// @smithy/core/dist-es/submodules/protocols/index.js
export { collectBody } from "./collect-stream-body";
export { extendedEncodeURIComponent } from "./extended-encode-uri-component";
export { HttpBindingProtocol } from "./HttpBindingProtocol";
export { HttpProtocol } from "./HttpProtocol";
export { RpcProtocol } from "./RpcProtocol";
export { RequestBuilder, requestBuilder } from "./requestBuilder";
export { resolvedPath } from "./resolve-path";
export { FromStringShapeDeserializer } from "./serde/FromStringShapeDeserializer";
export { HttpInterceptingShapeDeserializer } from "./serde/HttpInterceptingShapeDeserializer";
export { HttpInterceptingShapeSerializer } from "./serde/HttpInterceptingShapeSerializer";
export { ToStringShapeSerializer } from "./serde/ToStringShapeSerializer";
export { determineTimestampFormat } from "./serde/determineTimestampFormat";
export { SerdeContext } from "./SerdeContext";
export { Field } from "./protocol-http/Field";
export { Fields } from "./protocol-http/Fields";
export { HttpRequest } from "@smithy/core/transport";
export { HttpResponse } from "@smithy/core/transport";
export { isValidHostname } from "@smithy/core/transport";
export { getHttpHandlerExtensionConfiguration, resolveHttpHandlerRuntimeConfig, } from "./protocol-http/extensions/httpExtensionConfiguration";
export { contentLengthMiddleware, contentLengthMiddlewareOptions, getContentLengthPlugin, } from "./middleware-content-length/contentLengthMiddleware";
export { escapeUri } from "./util-uri-escape/escape-uri";
export { escapeUriPath } from "./util-uri-escape/escape-uri-path";
export { buildQueryString } from "./querystring-builder/buildQueryString";
export { parseQueryString } from "@smithy/core/transport";
export { parseUrl } from "@smithy/core/transport";
